File: backend/app/domain/entities/item.py
"""
Item Aggregate Root & Specialized Subtypes — مشروع «مُعين» (Mouin)
Item is the Aggregate Root. Types: task, appointment, note, document, debt, shopping.
Reminder is strictly decoupled from Item.
"""

from dataclasses import dataclass, field
from datetime import datetime, date, timezone
from typing import Optional
from backend.app.domain.entities.base import AggregateRoot
from backend.app.domain.value_objects.identity import EntityId, WorkspaceId, InstallationId
from backend.app.domain.value_objects.types import ItemType, PrivacyClassification, Priority, TaskStatus
from backend.app.domain.exceptions import InvariantViolationError, InvalidStateTransitionError
from backend.app.domain.events.domain_events import (
    ItemCreatedEvent, ItemUpdatedEvent, ItemSoftDeletedEvent, TaskCompletedEvent
)

@dataclass
class TaskDetail:
    due_date: Optional[datetime] = None
    priority: Priority = Priority.MEDIUM
    status: TaskStatus = TaskStatus.PENDING
    completed_at: Optional[datetime] = None
    estimated_duration_minutes: Optional[int] = None

    def __post_init__(self):
        if self.estimated_duration_minutes is not None and self.estimated_duration_minutes <= 0:
            raise InvariantViolationError("Estimated duration minutes must be positive.")

@dataclass
class AppointmentDetail:
    start_time: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    end_time: Optional[datetime] = None
    location: Optional[str] = None
    all_day: bool = False
    timezone: str = "Asia/Aden"

    def __post_init__(self):
        if self.end_time is not None and self.end_time < self.start_time:
            raise InvariantViolationError("Appointment end_time cannot be earlier than start_time.")

@dataclass
class NoteDetail:
    content: str = ""
    content_format: str = "plain_text"

    def __post_init__(self):
        if self.content_format not in ("plain_text", "markdown"):
            raise InvariantViolationError(f"Unsupported content format: {self.content_format}")

@dataclass
class DocumentDetail:
    document_type: str = ""
    issue_date: Optional[date] = None
    expiry_date: Optional[date] = None
    document_number: Optional[str] = None
    issuing_authority: Optional[str] = None

    def __post_init__(self):
        if not self.document_type or not self.document_type.strip():
            raise InvariantViolationError("Document document_type is required.")
        if self.expiry_date is not None and self.issue_date is not None:
            if self.expiry_date < self.issue_date:
                raise InvariantViolationError("Document expiry_date cannot be earlier than issue_date.")

@dataclass
class Item(AggregateRoot):
    item_type: ItemType = ItemType.TASK
    title: str = ""
    summary: Optional[str] = None
    category_id: Optional[EntityId] = None
    privacy_classification: PrivacyClassification = PrivacyClassification.PRIVATE
    temporal_original_expression: Optional[str] = None
    temporal_resolved_at: Optional[datetime] = None
    temporal_timezone: Optional[str] = None
    temporal_locale: Optional[str] = "ar"
    temporal_calendar: Optional[str] = "gregorian"
    created_by_installation_id: Optional[InstallationId] = None

    # Extension details (1:1 with Item)
    task_detail: Optional[TaskDetail] = None
    appointment_detail: Optional[AppointmentDetail] = None
    note_detail: Optional[NoteDetail] = None
    document_detail: Optional[DocumentDetail] = None

    def __post_init__(self):
        if not self.title or not self.title.strip():
            raise InvariantViolationError("Item title is mandatory and cannot be empty.")
        if str(self.item_type).lower() == "reminder":
            raise InvariantViolationError("Reminder is strictly an independent subsystem and CANNOT be an Item type.")
        if not isinstance(self.item_type, ItemType):
            try:
                self.item_type = ItemType(self.item_type)
            except ValueError:
                raise InvariantViolationError(f"Invalid item_type: {self.item_type}")

    @classmethod
    def create_task(
        cls,
        id: EntityId,
        workspace_id: WorkspaceId,
        title: str,
        due_date: Optional[datetime] = None,
        priority: Priority = Priority.MEDIUM,
        summary: Optional[str] = None,
        category_id: Optional[EntityId] = None,
        privacy: PrivacyClassification = PrivacyClassification.PRIVATE,
        temporal_expression: Optional[str] = None,
        installation_id: Optional[InstallationId] = None
    ) -> "Item":
        task_detail = TaskDetail(due_date=due_date, priority=priority, status=TaskStatus.PENDING)
        item = cls(
            id=id,
            workspace_id=workspace_id,
            item_type=ItemType.TASK,
            title=title,
            summary=summary,
            category_id=category_id,
            privacy_classification=privacy,
            temporal_original_expression=temporal_expression,
            temporal_resolved_at=due_date,
            created_by_installation_id=installation_id,
            task_detail=task_detail
        )
        item.record_event(ItemCreatedEvent(
            event_id=str(EntityId.new()),
            workspace_id=str(workspace_id),
            event_type="item.task_created",
            aggregate_type="item",
            aggregate_id=str(id),
            payload={"title": title, "priority": priority.value}
        ))
        return item

    def complete_task(self):
        if self.item_type != ItemType.TASK or not self.task_detail:
            raise InvalidStateTransitionError("Only Task items can be completed.")
        if self.task_detail.status == TaskStatus.COMPLETED:
            return  # Already completed
        self.task_detail.status = TaskStatus.COMPLETED
        self.task_detail.completed_at = datetime.now(timezone.utc)
        self.touch()
        self.record_event(TaskCompletedEvent(
            event_id=str(EntityId.new()),
            workspace_id=str(self.workspace_id),
            event_type="task.completed",
            aggregate_type="item",
            aggregate_id=str(self.id),
            payload={"completed_at": self.task_detail.completed_at.isoformat()}
        ))

    def update_title_and_summary(self, title: str, summary: Optional[str] = None):
        if not title or not title.strip():
            raise InvariantViolationError("Title cannot be empty.")
        self.title = title.strip()
        self.summary = summary
        self.touch()
        self.record_event(ItemUpdatedEvent(
            event_id=str(EntityId.new()),
            workspace_id=str(self.workspace_id),
            event_type="item.updated",
            aggregate_type="item",
            aggregate_id=str(self.id),
            payload={"title": self.title}
        ))

    def soft_delete(self):
        if not self.is_deleted():
            self.mark_deleted()
            self.record_event(ItemSoftDeletedEvent(
                event_id=str(EntityId.new()),
                workspace_id=str(self.workspace_id),
                event_type="item.deleted",
                aggregate_type="item",
                aggregate_id=str(self.id),
                payload={"deleted_at": self.deleted_at.isoformat()}
            ))

File: mobile/lib/domain/entities/item.dart
// Item Aggregate & Subtypes — مشروع «مُعين» (Mouin)
import '../value_objects/types.dart';

class TaskDetail {
  final Priority priority;
  final DateTime? dueDate;
  final TaskStatus status;
  final DateTime? completedAt;

  const TaskDetail({
    this.priority = Priority.medium,
    this.dueDate,
    this.status = TaskStatus.pending,
    this.completedAt,
  });

  TaskDetail copyWith({
    Priority? priority,
    DateTime? dueDate,
    TaskStatus? status,
    DateTime? completedAt,
  }) {
    return TaskDetail(
      priority: priority ?? this.priority,
      dueDate: dueDate ?? this.dueDate,
      status: status ?? this.status,
      completedAt: completedAt ?? this.completedAt,
    );
  }
}

class NoteDetail {
  final String content;
  final String contentFormat;

  const NoteDetail({
    required this.content,
    this.contentFormat = 'plain_text',
  });

  NoteDetail copyWith({
    String? content,
    String? contentFormat,
  }) {
    return NoteDetail(
      content: content ?? this.content,
      contentFormat: contentFormat ?? this.contentFormat,
    );
  }
}

class AppointmentDetail {
  final DateTime startTime;
  final DateTime? endTime;
  final String? location;
  final bool allDay;
  final String? timezone;

  const AppointmentDetail({
    required this.startTime,
    this.endTime,
    this.location,
    this.allDay = false,
    this.timezone,
  });

  AppointmentDetail copyWith({
    DateTime? startTime,
    DateTime? endTime,
    String? location,
    bool? allDay,
    String? timezone,
  }) {
    return AppointmentDetail(
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      location: location ?? this.location,
      allDay: allDay ?? this.allDay,
      timezone: timezone ?? this.timezone,
    );
  }
}

class DocumentDetail {
  final String? documentType;
  final String? documentNumber;
  final String? issuingAuthority;
  final DateTime? issueDate;
  final DateTime? expiryDate;

  const DocumentDetail({
    this.documentType,
    this.documentNumber,
    this.issuingAuthority,
    this.issueDate,
    this.expiryDate,
  });

  DocumentDetail copyWith({
    String? documentType,
    String? documentNumber,
    String? issuingAuthority,
    DateTime? issueDate,
    DateTime? expiryDate,
  }) {
    return DocumentDetail(
      documentType: documentType ?? this.documentType,
      documentNumber: documentNumber ?? this.documentNumber,
      issuingAuthority: issuingAuthority ?? this.issuingAuthority,
      issueDate: issueDate ?? this.issueDate,
      expiryDate: expiryDate ?? this.expiryDate,
    );
  }
}

class ShoppingListDetail {
  final List<dynamic> items;

  const ShoppingListDetail({
    this.items = const [],
  });

  ShoppingListDetail copyWith({
    List<dynamic>? items,
  }) {
    return ShoppingListDetail(
      items: items ?? this.items,
    );
  }
}

class Item {
  final String id;
  final String workspaceId;
  final ItemType itemType;
  final String title;
  final String? summary;
  final PrivacyClassification? privacy;
  final String? categoryId;
  final String? voiceFilePath;
  final int? voiceDurationMs;
  final TaskDetail? taskDetail;
  final NoteDetail? noteDetail;
  final AppointmentDetail? appointmentDetail;
  final DocumentDetail? documentDetail;
  final ShoppingListDetail? shoppingListDetail;
  final int entityVersion;
  final DateTime? deletedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  Item({
    required this.id,
    required this.workspaceId,
    required this.itemType,
    required this.title,
    this.summary,
    this.privacy,
    this.categoryId,
    this.voiceFilePath,
    this.voiceDurationMs,
    TaskDetail? taskDetail,
    Priority? priority,
    bool? isCompleted,
    DateTime? dueDate,
    this.noteDetail,
    this.appointmentDetail,
    this.documentDetail,
    this.shoppingListDetail,
    this.entityVersion = 1,
    this.deletedAt,
    required this.createdAt,
    required this.updatedAt,
  }) : taskDetail = taskDetail ??
            (itemType == ItemType.task || priority != null || isCompleted != null || dueDate != null
                ? TaskDetail(
                    priority: priority ?? Priority.medium,
                    dueDate: dueDate,
                    status: (isCompleted ?? false) ? TaskStatus.completed : TaskStatus.pending,
                  )
                : null);

  Priority get priority => taskDetail?.priority ?? Priority.medium;
  bool get isCompleted => taskDetail?.status == TaskStatus.completed;
  DateTime? get dueDate => taskDetail?.dueDate;
  bool get isDeleted => deletedAt != null;

  Item markDeleted({DateTime? at}) {
    final now = at ?? DateTime.now().toUtc();
    return copyWith(
      deletedAt: now,
      updatedAt: now,
      entityVersion: entityVersion + 1,
    );
  }

  factory Item.createTask({
    required String id,
    required String workspaceId,
    required String title,
    DateTime? dueDate,
    Priority priority = Priority.medium,
    String? summary,
    String? categoryId,
    PrivacyClassification? privacy,
  }) {
    final now = DateTime.now().toUtc();
    return Item(
      id: id,
      workspaceId: workspaceId,
      itemType: ItemType.task,
      title: title,
      summary: summary,
      categoryId: categoryId,
      privacy: privacy,
      taskDetail: TaskDetail(
        priority: priority,
        dueDate: dueDate,
        status: TaskStatus.pending,
      ),
      createdAt: now,
      updatedAt: now,
    );
  }

  factory Item.createNote({
    required String id,
    required String workspaceId,
    required String title,
    required String content,
    String contentFormat = 'plain_text',
    String? summary,
    String? categoryId,
    PrivacyClassification? privacy,
  }) {
    final now = DateTime.now().toUtc();
    return Item(
      id: id,
      workspaceId: workspaceId,
      itemType: ItemType.note,
      title: title,
      summary: summary,
      categoryId: categoryId,
      privacy: privacy,
      noteDetail: NoteDetail(
        content: content,
        contentFormat: contentFormat,
      ),
      createdAt: now,
      updatedAt: now,
    );
  }

  factory Item.createAppointment({
    required String id,
    required String workspaceId,
    required String title,
    required DateTime startTime,
    DateTime? endTime,
    String? location,
    bool allDay = false,
    String? timezone,
    String? summary,
    String? categoryId,
    PrivacyClassification? privacy,
  }) {
    final now = DateTime.now().toUtc();
    return Item(
      id: id,
      workspaceId: workspaceId,
      itemType: ItemType.appointment,
      title: title,
      summary: summary,
      categoryId: categoryId,
      privacy: privacy,
      appointmentDetail: AppointmentDetail(
        startTime: startTime,
        endTime: endTime,
        location: location,
        allDay: allDay,
        timezone: timezone,
      ),
      createdAt: now,
      updatedAt: now,
    );
  }

  factory Item.createDocument({
    required String id,
    required String workspaceId,
    required String title,
    String? documentType,
    String? documentNumber,
    String? issuingAuthority,
    DateTime? issueDate,
    DateTime? expiryDate,
    String? summary,
    String? categoryId,
    PrivacyClassification? privacy,
  }) {
    final now = DateTime.now().toUtc();
    return Item(
      id: id,
      workspaceId: workspaceId,
      itemType: ItemType.document,
      title: title,
      summary: summary,
      categoryId: categoryId,
      privacy: privacy,
      documentDetail: DocumentDetail(
        documentType: documentType,
        documentNumber: documentNumber,
        issuingAuthority: issuingAuthority,
        issueDate: issueDate,
        expiryDate: expiryDate,
      ),
      createdAt: now,
      updatedAt: now,
    );
  }

  factory Item.createUnified({
    required String id,
    required String workspaceId,
    required ItemType itemType,
    required String title,
    String? summary,
    String? categoryId,
    PrivacyClassification? privacy,
    TaskDetail? taskDetail,
    NoteDetail? noteDetail,
    AppointmentDetail? appointmentDetail,
    DocumentDetail? documentDetail,
    ShoppingListDetail? shoppingListDetail,
  }) {
    final now = DateTime.now().toUtc();
    return Item(
      id: id,
      workspaceId: workspaceId,
      itemType: itemType,
      title: title,
      summary: summary,
      categoryId: categoryId,
      privacy: privacy,
      taskDetail: taskDetail,
      noteDetail: noteDetail,
      appointmentDetail: appointmentDetail,
      documentDetail: documentDetail,
      shoppingListDetail: shoppingListDetail,
      createdAt: now,
      updatedAt: now,
    );
  }

  Item copyWith({
    String? title,
    String? summary,
    Priority? priority,
    bool? isCompleted,
    DateTime? dueDate,
    PrivacyClassification? privacy,
    String? categoryId,
    String? voiceFilePath,
    int? voiceDurationMs,
    TaskDetail? taskDetail,
    NoteDetail? noteDetail,
    AppointmentDetail? appointmentDetail,
    DocumentDetail? documentDetail,
    ShoppingListDetail? shoppingListDetail,
    int? entityVersion,
    DateTime? deletedAt,
    DateTime? updatedAt,
  }) {
    TaskDetail? updatedTaskDetail = taskDetail ?? this.taskDetail;
    if (priority != null || isCompleted != null || dueDate != null) {
      updatedTaskDetail = (updatedTaskDetail ?? const TaskDetail()).copyWith(
        priority: priority,
        dueDate: dueDate,
        status: isCompleted != null
            ? (isCompleted ? TaskStatus.completed : TaskStatus.pending)
            : updatedTaskDetail?.status,
      );
    }

    return Item(
      id: id,
      workspaceId: workspaceId,
      itemType: itemType,
      title: title ?? this.title,
      summary: summary ?? this.summary,
      privacy: privacy ?? this.privacy,
      categoryId: categoryId ?? this.categoryId,
      voiceFilePath: voiceFilePath ?? this.voiceFilePath,
      voiceDurationMs: voiceDurationMs ?? this.voiceDurationMs,
      taskDetail: updatedTaskDetail,
      noteDetail: noteDetail ?? this.noteDetail,
      appointmentDetail: appointmentDetail ?? this.appointmentDetail,
      documentDetail: documentDetail ?? this.documentDetail,
      shoppingListDetail: shoppingListDetail ?? this.shoppingListDetail,
      entityVersion: entityVersion ?? this.entityVersion,
      deletedAt: deletedAt ?? this.deletedAt,
      createdAt: createdAt,
      updatedAt: updatedAt ?? DateTime.now().toUtc(),
    );
  }
}
